File: backend/app/schemas/cases.py
"""ARCH43-S1:schemas-cases — case intelligence API models. The console mirrors
these in frontend/src/types/cases.ts; verify_arch43 compares the field lists."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field


class TemplateRow(BaseModel):
    id: uuid.UUID
    key: str
    version: int
    name: str
    description: str
    status: str
    assembly_key: str
    entity_kind: Optional[str] = None
    entity_role: Optional[str] = None
    required_documents: list[dict[str, Any]]
    rules: list[dict[str, Any]]
    request_ttl_hours: int
    published_at: Optional[datetime] = None


class TemplateWrite(BaseModel):
    key: str = Field(max_length=64)
    name: str = Field(max_length=160)
    description: str = Field(default="", max_length=4000)
    assembly_key: str = "MANUAL"
    entity_kind: Optional[str] = None
    entity_role: Optional[str] = None
    required_documents: list[dict[str, Any]] = Field(default_factory=list, max_length=40)
    rules: list[dict[str, Any]] = Field(default_factory=list, max_length=60)
    request_ttl_hours: int = 72


class CaseRow(BaseModel):
    id: uuid.UUID
    title: str
    status: str
    completeness: float
    template_id: uuid.UUID
    template_key: str
    anchor_kind: str
    anchor_entity_id: Optional[uuid.UUID] = None
    anchor_batch_id: Optional[uuid.UUID] = None
    documents: int
    failed_rules: int
    created_at: datetime
    evaluated_at: Optional[datetime] = None


class CaseList(BaseModel):
    items: list[CaseRow]
    counts_by_status: dict[str, int]


class ChecklistSlot(BaseModel):
    doc_type: str
    label: str
    min_count: int
    present: int
    satisfied: bool


class CaseDocumentRow(BaseModel):
    work_item_id: uuid.UUID
    original_filename: str
    document_type: str
    source: str
    added_at: datetime


class RuleResultRow(BaseModel):
    rule_id: str
    label: str
    op: str
    outcome: str
    left_value: Optional[str] = None
    right_value: Optional[str] = None
    detail: dict[str, Any]


class RequestRow(BaseModel):
    id: uuid.UUID
    document_type: str
    recipient_label: str
    status: str
    expires_at: datetime
    used_at: Optional[datetime] = None
    fulfilled_work_item_id: Optional[uuid.UUID] = None


class CaseDetail(BaseModel):
    case: CaseRow
    template: TemplateRow
    checklist: list[ChecklistSlot]
    documents: list[CaseDocumentRow]
    rules: list[RuleResultRow]
    requests: list[RequestRow]


class CaseCreate(BaseModel):
    template_id: uuid.UUID
    title: str = Field(max_length=200)


class CaseDocumentAdd(BaseModel):
    work_item_id: uuid.UUID
    document_type: Optional[str] = Field(default=None, max_length=64)


class RequestCreate(BaseModel):
    document_type: str = Field(max_length=64)
    recipient_label: str = Field(default="", max_length=200)
    ttl_hours: Optional[int] = None


class RequestCreated(BaseModel):
    request: RequestRow
    token: str
    upload_path: str


class PublicRequestInfo(BaseModel):
    document_type: str
    case_title: str
    expires_at: datetime


class PublicUploadResult(BaseModel):
    received: bool
    document_type: str

File: frontend/src/types/cases.ts
/**
 * ARCH43-S2:types-cases — case intelligence API shapes. Mirrors
 * backend/app/schemas/cases.py field for field (verify_arch43 compares them).
 */

export type CaseStatus = "INCOMPLETE" | "INCONSISTENT" | "COMPLETE" | "CLOSED";
export type RuleOutcome = "PASS" | "FAIL" | "MISSING" | "ERROR";
export const CASE_STATUSES: readonly CaseStatus[] = ["INCOMPLETE", "INCONSISTENT", "COMPLETE", "CLOSED"];
export const CASE_STATUS_LABELS: Readonly<Record<CaseStatus, string>> = {
  INCOMPLETE: "Waiting for documents",
  INCONSISTENT: "Inconsistent",
  COMPLETE: "Complete",
  CLOSED: "Closed",
};

export interface TemplateRow {
  readonly id: string;
  readonly key: string;
  readonly version: number;
  readonly name: string;
  readonly description: string;
  readonly status: "DRAFT" | "PUBLISHED" | "RETIRED";
  readonly assembly_key: "ENTITY" | "BATCH" | "MANUAL";
  readonly entity_kind: string | null;
  readonly entity_role: string | null;
  readonly required_documents: readonly Readonly<Record<string, unknown>>[];
  readonly rules: readonly Readonly<Record<string, unknown>>[];
  readonly request_ttl_hours: number;
  readonly published_at: string | null;
}

export interface TemplateWrite {
  readonly key: string;
  readonly name: string;
  readonly description?: string;
  readonly assembly_key: string;
  readonly entity_kind?: string | null;
  readonly entity_role?: string | null;
  readonly required_documents: readonly Readonly<Record<string, unknown>>[];
  readonly rules: readonly Readonly<Record<string, unknown>>[];
  readonly request_ttl_hours?: number;
}

export interface CaseRow {
  readonly id: string;
  readonly title: string;
  readonly status: CaseStatus;
  readonly completeness: number;
  readonly template_id: string;
  readonly template_key: string;
  readonly anchor_kind: string;
  readonly anchor_entity_id: string | null;
  readonly anchor_batch_id: string | null;
  readonly documents: number;
  readonly failed_rules: number;
  readonly created_at: string;
  readonly evaluated_at: string | null;
}

export interface CaseList {
  readonly items: readonly CaseRow[];
  readonly counts_by_status: Readonly<Record<string, number>>;
}

export interface ChecklistSlot {
  readonly doc_type: string;
  readonly label: string;
  readonly min_count: number;
  readonly present: number;
  readonly satisfied: boolean;
}

export interface CaseDocumentRow {
  readonly work_item_id: string;
  readonly original_filename: string;
  readonly document_type: string;
  readonly source: string;
  readonly added_at: string;
}

export interface RuleResultRow {
  readonly rule_id: string;
  readonly label: string;
  readonly op: string;
  readonly outcome: RuleOutcome;
  readonly left_value: string | null;
  readonly right_value: string | null;
  readonly detail: Readonly<Record<string, unknown>>;
}

export interface RequestRow {
  readonly id: string;
  readonly document_type: string;
  readonly recipient_label: string;
  readonly status: "OPEN" | "FULFILLED" | "EXPIRED" | "REVOKED";
  readonly expires_at: string;
  readonly used_at: string | null;
  readonly fulfilled_work_item_id: string | null;
}

export interface CaseDetail {
  readonly case: CaseRow;
  readonly template: TemplateRow;
  readonly checklist: readonly ChecklistSlot[];
  readonly documents: readonly CaseDocumentRow[];
  readonly rules: readonly RuleResultRow[];
  readonly requests: readonly RequestRow[];
}

export interface RequestCreated {
  readonly request: RequestRow;
  readonly token: string;
  readonly upload_path: string;
}

export interface PublicRequestInfo {
  readonly document_type: string;
  readonly case_title: string;
  readonly expires_at: string;
}

export interface PublicUploadResult {
  readonly received: boolean;
  readonly document_type: string;
}
